File: scrape-images.js
import https from 'https';
import http from 'http';
import { createWriteStream, mkdirSync, existsSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { parse } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const TARGET_URL = 'https://www.balloonlightprag.cz';
const OUTPUT_DIR = join(__dirname, 'public', 'images');

// Create output directory if it doesn't exist
if (!existsSync(OUTPUT_DIR)) {
  mkdirSync(OUTPUT_DIR, { recursive: true });
  console.log(`Created directory: ${OUTPUT_DIR}`);
}

async function fetchHTML(url) {
  return new Promise((resolve, reject) => {
    const parsedUrl = parse(url);
    const client = parsedUrl.protocol === 'https:' ? https : http;
    
    client.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
      }
    }, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

function extractImageUrls(html, baseUrl) {
  const imageUrls = new Set();
  
  // Extract src attributes from img tags (more flexible regex)
  const imgRegex = /<img[^>]+(?:src|srcset|data-src|data-lazy-src|data-srcset)=["']([^"']+)["']/gi;
  let match;
  while ((match = imgRegex.exec(html)) !== null) {
    let imageUrl = match[1];
    
    // Handle srcset (might have multiple URLs)
    if (imageUrl.includes(',')) {
      const urls = imageUrl.split(',').map(u => u.trim().split(/\s+/)[0]);
      urls.forEach(url => {
        let normalized = normalizeUrl(url, baseUrl);
        if (normalized && isValidImageUrl(normalized)) {
          imageUrls.add(normalized);
        }
      });
      continue;
    }
    
    let normalized = normalizeUrl(imageUrl, baseUrl);
    if (normalized && isValidImageUrl(normalized)) {
      imageUrls.add(normalized);
    }
  }
  
  // Also check for background-image in style attributes
  const styleRegex = /style=["'][^"']*background(?:-image)?:\s*url\(["']?([^"')]+)["']?\)/gi;
  while ((match = styleRegex.exec(html)) !== null) {
    let imageUrl = match[1];
    let normalized = normalizeUrl(imageUrl, baseUrl);
    if (normalized && isValidImageUrl(normalized)) {
      imageUrls.add(normalized);
    }
  }
  
  // Check for image URLs in JSON data (common in React/Vue apps)
  const jsonRegex = /"(?:https?:\/\/[^"]+\.(?:jpg|jpeg|png|gif|webp|svg)[^"]*)"/gi;
  while ((match = jsonRegex.exec(html)) !== null) {
    let imageUrl = match[1] || match[0].replace(/"/g, '');
    let normalized = normalizeUrl(imageUrl, baseUrl);
    if (normalized && isValidImageUrl(normalized)) {
      imageUrls.add(normalized);
    }
  }
  
  // Look for common image paths
  const pathRegex = /(?:src|href|data-src)=["']([^"']*(?:images?|img|photo|gallery|assets)[^"']*\.(?:jpg|jpeg|png|gif|webp|svg)[^"']*)["']/gi;
  while ((match = pathRegex.exec(html)) !== null) {
    let imageUrl = match[1];
    let normalized = normalizeUrl(imageUrl, baseUrl);
    if (normalized && isValidImageUrl(normalized)) {
      imageUrls.add(normalized);
    }
  }
  
  return Array.from(imageUrls);
}

function normalizeUrl(url, baseUrl) {
  if (!url || url.startsWith('data:')) return null;
  
  // Convert relative URLs to absolute
  if (url.startsWith('//')) {
    return 'https:' + url;
  } else if (url.startsWith('/')) {
    return new URL(url, baseUrl).href;
  } else if (!url.startsWith('http')) {
    return new URL(url, baseUrl).href;
  }
  
  return url;
}

function isValidImageUrl(url) {
  if (!url) return false;
  // Filter out icons, logos, and very small images, but include everything else
  const lowerUrl = url.toLowerCase();
  if (lowerUrl.includes('icon') || lowerUrl.includes('favicon')) {
    // Still include if it's a photo/image file
    if (!lowerUrl.match(/\.(jpg|jpeg|png|gif|webp)$/)) {
      return false;
    }
  }
  return true;
}

async function downloadImage(url, outputPath) {
  return new Promise((resolve, reject) => {
    const parsedUrl = parse(url);
    const client = parsedUrl.protocol === 'https:' ? https : http;
    
    client.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      }
    }, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        // Follow redirects
        return downloadImage(res.headers.location, outputPath)
          .then(resolve)
          .catch(reject);
      }
      
      if (res.statusCode !== 200) {
        return reject(new Error(`Failed to download: ${url} - Status: ${res.statusCode}`));
      }
      
      const fileStream = createWriteStream(outputPath);
      res.pipe(fileStream);
      
      fileStream.on('finish', () => {
        fileStream.close();
        resolve(outputPath);
      });
      
      fileStream.on('error', reject);
    }).on('error', reject);
  });
}

function getFilenameFromUrl(url) {
  try {
    const parsed = parse(url);
    const pathname = parsed.pathname || '';
    const filename = pathname.split('/').pop() || 'image';
    
    // Ensure filename has an extension
    if (!filename.match(/\.(jpg|jpeg|png|gif|webp|svg)$/i)) {
      return filename + '.jpg';
    }
    
    // Sanitize filename
    return filename.replace(/[^a-zA-Z0-9._-]/g, '_');
  } catch (e) {
    return 'image_' + Date.now() + '.jpg';
  }
}

async function main() {
  console.log(`Scraping images from: ${TARGET_URL}\n`);
  
  try {
    // Fetch the main page
    console.log('Fetching HTML...');
    const html = await fetchHTML(TARGET_URL);
    console.log('HTML fetched successfully.\n');
    
    // Save HTML for debugging
    writeFileSync(join(__dirname, 'debug-html.html'), html);
    console.log('Saved HTML to debug-html.html for inspection.\n');
    
    // Extract image URLs
    console.log('Extracting image URLs...');
    const imageUrls = extractImageUrls(html, TARGET_URL);
    console.log(`Found ${imageUrls.length} unique images.`);
    if (imageUrls.length > 0) {
      console.log('Sample URLs:');
      imageUrls.slice(0, 5).forEach((url, i) => console.log(`  ${i + 1}. ${url}`));
    }
    console.log('');
    
    // Download each image
    console.log('Downloading images...\n');
    let successCount = 0;
    let failCount = 0;
    
    for (let i = 0; i < imageUrls.length; i++) {
      const url = imageUrls[i];
      const filename = getFilenameFromUrl(url);
      const outputPath = join(OUTPUT_DIR, filename);
      
      // Skip if file already exists
      if (existsSync(outputPath)) {
        console.log(`[${i + 1}/${imageUrls.length}] Skipped (exists): ${filename}`);
        successCount++;
        continue;
      }
      
      try {
        await downloadImage(url, outputPath);
        console.log(`[${i + 1}/${imageUrls.length}] Downloaded: ${filename}`);
        successCount++;
        
        // Small delay to be polite
        await new Promise(resolve => setTimeout(resolve, 200));
      } catch (error) {
        console.error(`[${i + 1}/${imageUrls.length}] Failed: ${filename} - ${error.message}`);
        failCount++;
      }
    }
    
    console.log(`\n=== Summary ===`);
    console.log(`Total images found: ${imageUrls.length}`);
    console.log(`Successfully downloaded: ${successCount}`);
    console.log(`Failed: ${failCount}`);
    console.log(`\nImages saved to: ${OUTPUT_DIR}`);
    
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

main();

File: scrape-images.py
#!/usr/bin/env python3
import requests
from bs4 import BeautifulSoup
import os
from urllib.parse import urljoin, urlparse
import re
from pathlib import Path

TARGET_URL = 'https://www.balloonlightprag.cz'
OUTPUT_DIR = Path('public/images')
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

def fetch_html(url):
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    }
    response = requests.get(url, headers=headers, timeout=30)
    response.raise_for_status()
    return response.text

def extract_image_urls(html, base_url):
    soup = BeautifulSoup(html, 'html.parser')
    image_urls = set()
    
    # Find all img tags
    for img in soup.find_all('img'):
        for attr in ['src', 'srcset', 'data-src', 'data-lazy-src', 'data-srcset']:
            if img.get(attr):
                urls = img.get(attr).split(',')
                for url in urls:
                    url = url.strip().split()[0]  # Handle srcset format
                    if url and not url.startswith('data:'):
                        full_url = urljoin(base_url, url)
                        if is_image_url(full_url):
                            image_urls.add(full_url)
    
    # Find images in style attributes
    for elem in soup.find_all(style=True):
        style = elem['style']
        bg_urls = re.findall(r'url\(["\']?([^"\')]+)["\']?\)', style)
        for url in bg_urls:
            if not url.startswith('data:'):
                full_url = urljoin(base_url, url)
                if is_image_url(full_url):
                    image_urls.add(full_url)
    
    # Find images in links (common for galleries)
    for link in soup.find_all('a', href=True):
        href = link['href']
        if re.search(r'\.(jpg|jpeg|png|gif|webp|svg)$', href, re.IGNORECASE):
            full_url = urljoin(base_url, href)
            image_urls.add(full_url)
    
    # Look for images in script tags (JSON data)
    for script in soup.find_all('script'):
        if script.string:
            json_urls = re.findall(r'"(https?://[^"]+\.(?:jpg|jpeg|png|gif|webp|svg)[^"]*)"', script.string, re.IGNORECASE)
            image_urls.update(json_urls)
    
    return list(image_urls)

def is_image_url(url):
    if not url:
        return False
    url_lower = url.lower()
    # Exclude favicons and small icons unless they're actual image files
    if 'favicon' in url_lower or ('icon' in url_lower and not re.search(r'\.(jpg|jpeg|png|gif|webp)$', url_lower)):
        return False
    return True

def get_filename_from_url(url):
    parsed = urlparse(url)
    filename = os.path.basename(parsed.path) or 'image.jpg'
    if not re.search(r'\.(jpg|jpeg|png|gif|webp|svg)$', filename, re.IGNORECASE):
        filename += '.jpg'
    # Sanitize filename
    filename = re.sub(r'[^a-zA-Z0-9._-]', '_', filename)
    return filename

def download_image(url, output_path):
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    response = requests.get(url, headers=headers, timeout=30, stream=True)
    response.raise_for_status()
    
    with open(output_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)
    
    return output_path

def main():
    print(f'Scraping images from: {TARGET_URL}\n')
    
    try:
        # Fetch HTML
        print('Fetching HTML...')
        html = fetch_html(TARGET_URL)
        print('HTML fetched successfully.\n')
        
        # Save HTML for debugging
        with open('debug-html.html', 'w', encoding='utf-8') as f:
            f.write(html)
        print('Saved HTML to debug-html.html for inspection.\n')
        
        # Extract image URLs
        print('Extracting image URLs...')
        image_urls = extract_image_urls(html, TARGET_URL)
        print(f'Found {len(image_urls)} unique images.')
        if image_urls:
            print('Sample URLs:')
            for i, url in enumerate(image_urls[:5], 1):
                print(f'  {i}. {url}')
        print()
        
        # Download images
        print('Downloading images...\n')
        success_count = 0
        fail_count = 0
        
        for i, url in enumerate(image_urls, 1):
            filename = get_filename_from_url(url)
            output_path = OUTPUT_DIR / filename
            
            # Skip if exists
            if output_path.exists():
                print(f'[{i}/{len(image_urls)}] Skipped (exists): {filename}')
                success_count += 1
                continue
            
            try:
                download_image(url, output_path)
                print(f'[{i}/{len(image_urls)}] Downloaded: {filename}')
                success_count += 1
            except Exception as e:
                print(f'[{i}/{len(image_urls)}] Failed: {filename} - {str(e)}')
                fail_count += 1
        
        print(f'\n=== Summary ===')
        print(f'Total images found: {len(image_urls)}')
        print(f'Successfully downloaded: {success_count}')
        print(f'Failed: {fail_count}')
        print(f'\nImages saved to: {OUTPUT_DIR.absolute()}')
        
    except Exception as e:
        print(f'Error: {str(e)}')
        import traceback
        traceback.print_exc()
        exit(1)

if __name__ == '__main__':
    main()

